contracts/core/proofAPI/CrossL2EventProver.sol

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+ * Copyright 2024, Polymer Labs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+pragma solidity ^0.8.0;
+
+import {RLPReader} from "optimism/libraries/rlp/RLPReader.sol";
+import {MerkleTrie} from "optimism/libraries/trie/MerkleTrie.sol";
+import {Bytes} from "optimism/libraries/Bytes.sol";
+import {PolymerProofs} from "../../libs/PolymerProofs.sol";
+import {AppStateVerifier} from "../../base/AppStateVerifier.sol";
+import {ICrossEventProver} from "../../interfaces/ICrossEventProver.sol";
+import {Ics23Proof} from "../../interfaces/IProofVerifier.sol";
+import {ISignatureVerifier} from "../../interfaces/ISignatureVerifier.sol";
+import {LightClientType} from "../../interfaces/ILightClient.sol";
+
+contract CrossL2EventProver is AppStateVerifier, ICrossEventProver {
+    LightClientType public constant LIGHT_CLIENT_TYPE = LightClientType.SequencerLightClient; // Stored as a constant
+        // for cheap on-chain use
+
+    ISignatureVerifier public immutable verifier;
+
+    // Store peptide apphashes for a given height
+    mapping(uint256 => uint256) public peptideAppHashes;
+
+    error CannotUpdateClientWithDifferentAppHash();
+
+    constructor(ISignatureVerifier verifier_) {
+        verifier = verifier_;
+    }
+
+    /**
+     * @inheritdoc ICrossEventProver
+     */
+    function validateReceipt(bytes32 receiptRoot, bytes calldata eventProof) external view returns (bool valid) {
+        (
+            Heights memory heights,
+            bytes memory rawReceipt,
+            Ics23Proof memory peptideAppProof,
+            bytes[] memory receiptMMPTProof,
+            bytes32 peptideClientId
+        ) = abi.decode(eventProof, (Heights, bytes, Ics23Proof, bytes[], bytes32));
+        // Before we can trust the receipt root, we first need to verify that the receipt root is indeed stored
+        // on peptide at the given clientID and height.
+
+        // VerifyMembership verifies the receipt root  through an ics23 proof of peptide state that attests that the
+        // given sourceL2Height has the receipt root at the peptide height
+        this.verifyMembership(
+            bytes32(_getPeptideAppHash(peptideAppProof.height)),
+            PolymerProofs.receiptProofKey(peptideClientId, heights.sourceL2Height),
```

Comment:
verifier contract needs to check client id.

Reply:
peptide client id should be the hardcoded one per contract